gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -166,11 +166,22 @@ def self._on_method_added(hook_mod, method_name, is_singleton_method: false)
 
     current_declaration = T::Private::DeclState.current.active_declaration
     mod = is_singleton_method ? hook_mod.singleton_class : hook_mod
+    current_mod_final = T::Private::Final.final_module?(mod)
     original_method = mod.instance_method(method_name)
 
+    if !current_declaration.nil? && current_declaration.final && current_mod_final
+      raise "`#{mod.name}` was declared as final and its method `#{method_name}` " \
+        "cannot also be explicitly declared as final (it is implicitly final)"
+    end
```

Comment:
I would maybe err on the other side (must redeclare final). That makes it a little easier to grep for, for example, and should make the statics easier (because the sig parsing logic works the same way in either case).

Reply:
In that case, there's no particular benefit to declaring a module (which isn't a class) as final, since you'll have to mark every method in the module as final anyway, and you don't get any extra new stuff by declaring a regular module as final (as opposed to declaring a class final, which means that you now know no one subclasses you).

This would also affect users of `T::Struct` if we were to subsume the one-off logic that forbids inheriting from subclasses of `T::Struct` with just marking the subclasses as final.

Not to mention I foresee user complaints about how we could just "do it for you" and "why do I have to be explicit", etc.

I agree with the benefits you propose and would be fine either way. Actually, with me being the implementor, and a mentioned benefit of requiring the redeclaration being that the implementation becomes easier, I might be inclined to lean towards requiring redeclaring. I just want to make sure we're considering the effects of either option.

Reply:
Resolved: we will require marking methods as final, too.